### PR TITLE
[WIP] Syft 0.4 Prototype

### PR DESCRIPTION
**Pull request by iamtrask**
This is a project of the Syft Core team.

This PR focuses on initial support for type checking, custom constructors, and custom IDs for objects.

Development roadmap can be found at https://github.com/OpenMined/PySyft/blob/syft_4v2/experimental/trask/Syft%200.4%20Development%20Roadmap.ipynb

All development is being recorded via screenshare which is available upon request (might go on youtube later)

### CODE REVIEW

File: .idea/PySyft.iml [hunk 1]
```diff
@@ -0,0 +1,15 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (LaRiffle) [inline .idea/PySyft.iml]: Maybe rm .idea from your git?

**Comment by iamtrask** [comment]
It won’t be merged for a long time - once we’ve pulled in a lot of functionality from 0.3. That metric will probably still be quite high because of major code reorganisation.

File: syft/util.py [hunk 2]
```diff
@@ -0,0 +1,78 @@
+from syft import check
+import inspect
+
+
+@check.type_hints
+def is_static_method(klass: type, attr: str):
```
Comment (karlhigley) [inline syft/util.py]: According to [PEP 8](https://www.python.org/dev/peps/pep-0008/#descriptive-naming-styles), the convention for avoiding naming conflicts with reserved keywords is to use a single trailing underscore (like `class_`.)

File: syft/util.py [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+
+    if not inspect.isclass(klass):
+        return False
+
+    if hasattr(klass, attr):
+        value = getattr(klass, attr)
+    else:
+        return False
+
+    assert getattr(klass, attr) == value
```
Comment (karlhigley) [inline syft/util.py]: What is this `assert` doing exactly? Checking that the class method is idempotent? Can it be written without `assert`?

File: syft/util.py [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+                    return True
+    return False
+
+
+@check.type_hints
+def copy_static_methods(from_class: type, to_class: type):
+    """Copies all static methods from one class to another class
+
+    This utility was initialized during the creation of the Constructor for PyTorch's "th.Tensor" class. Since we
+    replace each original constructor (th.Tensor) with on we implement (torch_.UppercaseTensorConstructor), we also
```
Comment (karlhigley) [inline syft/util.py]:
```suggestion
    replace each original constructor (th.Tensor) with one we implement (torch_.UppercaseTensorConstructor), we also
```

File: syft/util.py [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+    need to make sure that our new constructor has any static methods which were previously stored on th.Tensor.
+    Otherwise, the library might look for them there, not find them, and then trigger an error.
+
+    Args:
+        from_class (Type): the class on which we look for static methods co copy
+        to_class (Type): the class onto which we copy all static methods found in <from_class>
+
+    """
+    # there are no static methods if from_class itself is not a type (sometimes funcs get passed in)
```
Comment (karlhigley) [inline syft/util.py]: Can we check that and return immediately if `from_class` isn't a class?

File: syft/util.py [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+
+@check.type_hints
+def get_original_constructor_name(object_name: str):
+    """Generate name for original constructor
+
+    For each custom constructor, we move the original constructor to a consistent location relative to
+    the original constructor so that each custom constructor automatically knows where to find the original
+     method it is overloading. Namely, we move the original constructor to a different attr within the same
+      module as the original constructor. This method specifies the naming convention that we use to name
+      the original constructor when it is moved.
```
Comment (karlhigley) [inline syft/util.py]:
```suggestion
     method it is overloading. Namely, we move the original constructor to a different attr within the same
     module as the original constructor. This method specifies the naming convention that we use to name
     the original constructor when it is moved.
```

File: test/torch_/test_native.py [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+
+
+def test_lowercase_tensor_constructor():
+    x = th.tensor([1, 2, 3, 4])
+
+    assert isinstance(x, th.Tensor)
+    assert (x == th.Tensor([1, 2, 3, 4])).all()
+
+
+def test_uppercase_tensor_constructor():
```
Comment (karlhigley) [inline test/torch_/test_native.py]: What's the difference between upper and lower case tensors? 🤔

File: syft/numpy_/array.py [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+    constructor_name = "array"
+
+    # Step 2: Store a reference to the location on which this constructor currently lives.
+    # This is also the location that this custom constructor will live once installed using
+    # self.install_inside_library()
+    constructor_location = np
+
+
+# Step 3: create constructor and install it in the library
+ArrayConstructor().install_inside_library()
```
[the remark this answers is not in the document]
Comment (karlhigley) [reply syft/numpy_/array.py]: This is better:
```
  ArrayConstructor(np).install_into_module()
```

**Comment by iamtrask** [comment]
> There are 476 file changes? Is this right? 😮

I don't mess around

**Comment by mari-linhares** [comment]
now it's only 474 file changes, I guess now I can have a look :)

**Comment by karlhigley** [comment]
"Hide deleted files" in the file filter is definitely your friend on this one. 😄

File: syft/check.py [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+    #   As a convention across PySyft,
+    #
+    #   Arg:
+    #       decorated (func): the function on which we want to check argument types
+    #
+    #   Returns:
+    #       decorator (func): the decorated function which checks argument types
+    #   """
+
+    def check_argument_types(args):
```
Comment (mari-linhares) [inline syft/check.py]: Why not use http://mypy-lang.org/?

I'm not familiar with it but it seems to do the job.

**Comment by mari-linhares** [comment]
@iamtrask this was marked as done but the PR wasn't merged. Have you moved this to a different PR?

**Comment by karlhigley** [comment]
This PR sat dormant/abandoned for a long time, but a lot of it is now in #3559 (which I just opened today.)

**Pull request closed without merging**